rust/datafusion/src/execution/physical_plan/csv.rs

```diff
@@ -71,15 +75,35 @@ impl CsvExec {
     /// Create a new execution plan for reading a set of CSV files
     pub fn try_new(
         path: &str,
-        schema: Arc<Schema>,
+        schema: Option<Arc<Schema>>,
         has_header: bool,
+        delimiter: Option<u8>,
         projection: Option<Vec<usize>>,
         batch_size: usize,
     ) -> Result<Self> {
+        let schema = match schema {
+            Some(s) => s,
+            None => {
+                let mut filenames: Vec<String> = vec![];
+                common::build_file_list(path, &mut filenames, ".csv")?;
+                if filenames.is_empty() {
+                    return Err(ExecutionError::General("No files found".to_string()));
+                }
+
+                let f = File::open(&filenames[0])?;
```

Comment:
What would happen if the first file is empty, but other files aren't? A good follow-up could be iterating through the files until an `Option<max_inference: usize>` is reached to ensure that the desired sample size is reached.

I've personally encountered some CSV files where the first k records conformed to a schema, but things broke down thereafter, especially with files that aren't generated by a uniformity-preserving tool like Spark or a database.

Reply:
yeah, there is no guarantee no matter what we do unless we read all the entries. Even with max_inference, we can't guarantee the remaining rows will confirm to the inferred schema.

the way i look at this is manually specify a schema if you want correctness and performance. only use schema inference if you just want to get a quick and dirty query up and running.

that said, i will try to change it to read all files instead of first one.

Reply:
I hear your argument, maybe let's wait for Andy's opinion first

Reply:
Yes, we could definitely make this more sophisticated (as a follow up PR). I think this is great for examples and getting started though.